### ISSUE TITLE
file400.error: File initialization failed.

### ISSUE THREAD

**Issue by jkyeung** [perg/file400#1]
I realize this isn't very much information to go on, but I am getting this error no matter what I try.

When I restored the SAVF, I did get some notice about QDFTOWN. I tried changing the *PUBLIC authority to *ALL for both the PYTHON3 library and the RECLEVACC service program, which didn't seem to make a difference.

I am happy to provide more details if you tell me what to look for.

**Comment by perg** [comment]
It may be an authority issue.You could try to change the owner of the srvpgm be a user with full authority to the database and set usrprf to *owner, chgsrvpgm python3/reclevacc usrprf(*owner)

**Comment by jkyeung** [comment]
> It may be an authority issue.You could try to change the owner of the srvpgm be a user with full authority to the database and set usrprf to *owner, chgsrvpgm python3/reclevacc usrprf(*owner)

I just tried those things and am still getting the same error.

Could the system CCSID be an issue at all? We are one of those old-school shops that uses 65535.

**Comment by perg** [comment]
I'm not sure, we don't use 65535.
The lib/file name is converted from ascii to the job ccsid, using QtqIconvOpen. If the CCSID of the current job is 65535, the default CCSID from the DFTCCSID job attribute is used.

**Comment by jkyeung** [comment]
> The lib/file name is converted from ascii to the job ccsid, using QtqIconvOpen. If the CCSID of the current job is 65535, the default CCSID from the DFTCCSID job attribute is used.

That makes sense. As far as I can tell, our jobs have DFTCCSID=37, so it doesn't seem like the problem is related to text encoding.

**Comment by kadler** [comment]
Searching through the code in GitHub (which GH search can be hit or miss), there's only one point which returns that exact message: https://github.com/perg/file400/blob/02aed765c9fdd20bd57497ab6223966b5c3f3d75/source/file400.c#L2589-L2593

This calls [call_fileGetStruct](https://github.com/perg/file400/blob/02aed765c9fdd20bd57497ab6223966b5c3f3d75/source/file400.c#L581-L590) which calls the ILE [fileGetStruct](https://github.com/perg/file400/blob/master/source/reclevacc.c#L471-L484) function.

Of course that only happens if [call_fileInit](https://github.com/perg/file400/blob/02aed765c9fdd20bd57497ab6223966b5c3f3d75/source/file400.c#L360-L368) returns 0 and fileGetStruct doesn't seem like it returns anything other than 0, so it is likely in [fileInit](https://github.com/perg/file400/blob/master/source/reclevacc.c#L291) so it looks like the problem is from one of the calls to QDBRTVFD.

**Comment by jkyeung** [comment]
> [...] so [the failure] is likely in [fileInit](https://github.com/perg/file400/blob/master/source/reclevacc.c#L291) so it looks like the problem is from one of the calls to QDBRTVFD.

Indeed. Through experimentation, it seemed that I didn't have (much) problem with files that only have a smallish number of fields. The file that was giving me problems has over 300 fields. I found [this thread](https://code400.com/forum/forum/iseries-programming-languages/rpg-rpgle/13076-limit-in-qdbrtvfd-api) talking about a similar issue. If it's true that the FILD0200 format can return over 400 bytes per field, then maybe it's exceeding the [allocated buffer](https://github.com/perg/file400/blob/master/source/reclevacc.c#L311-L318) (100000 bytes)?

**Comment by darrenspielman** [comment]
Getting this error.. any thoughts? allocated buffer should not be exceeding the max.
